libp2p/protocols/pubsub/pubsub.nim

```diff
@@ -581,37 +586,46 @@ method removeValidator*(
 
 method validate*(
     p: PubSub, message: Message
-): Future[ValidationResult] {.async, base.} =
+): Future[ValidationResult] {.async: (raises:[CancelledError]), base.} =
   var pending: seq[Future[ValidationResult]]
   trace "about to validate message"
   let topic = message.topic
   trace "looking for validators on topic",
     topic = topic, registered = toSeq(p.validators.keys)
   if topic in p.validators:
     trace "running validators for topic", topic = topic
-    for validator in p.validators[topic]:
-      pending.add(validator(topic, message))
+    try:
+      for validator in p.validators[topic]:
+        pending.add(validator(topic, message))
+    except KeyError:
+      raiseAssert "topic checked before"
 
-  result = ValidationResult.Accept
+  var valResult = ValidationResult.Accept
   let futs = await allFinished(pending)
   for fut in futs:
     if fut.failed:
-      result = ValidationResult.Reject
+      valResult = ValidationResult.Reject
       break
-    let res = fut.read()
-    if res != ValidationResult.Accept:
-      result = res
-      if res == ValidationResult.Reject:
-        break
-
-  case result
+    try:
+      let res = fut.read()
+      if res != ValidationResult.Accept:
+        valResult = res
+        if res == ValidationResult.Reject:
+          break
+    except CatchableError as e:
+      trace "validator for message could not be executed, ignoring", topic = topic, err = e.msg
+      valResult = ValidationResult.Ignore 
```

Comment:
To avoid returning a `CatchableError` I capture the exception here, and ignore the message.